devito/types/basic.py

```diff
@@ -770,6 +770,15 @@ def evaluate(self):
             return self
         return weight * sum(avg_list)
 
+    def _subs(self, old, new, **hints):
```

Comment:
this looks identical to the one in Differentiable? Since DiscreteFunction is a subclass of both AbstractFunction and Differentiable, why don't we just drop it?

Reply:
This is not in Differentiable, only in DifferentiableOp.

Reply:
Also previously missed small cases that can happen only for a function (or symbol) where you can have `u._subs(u, v)` which is not the case for an expression.

Reply:
I'm not sure I follow that last past... the `if old is self` at least should apply to generic exprs too?

I have a feeling, potentially wrong, that there's something off with this _subs being different than what we got in DifferentiableOp.

Other question: why is `Differentiable.subs` different than `DifferentiableOp.subs`? Should we rather try to have just one general `_subs` in `Differentiable`?

Reply:
> just one general _subs in Differentiable?

Way slower (no idea why) and also does weird stuff for non Add/Mul so why was originally moved to DifferentialOp

Reply:
so why the below 

```
        if old is self:
             return new
         if old is new:
             return self
```

doesn't apply to DifferentiableOp?